server-common/src/main/java/org/apache/kafka/server/metrics/KafkaMetricsGroup.java

```diff
@@ -0,0 +1,161 @@
+            nameBuilder.append(name);
+        }
+
+        final String scope = toScope(tags).orElse(null);
+        final Optional<String> tagsName = toMBeanName(tags);
+        tagsName.ifPresent(s -> nameBuilder.append(",").append(s));
+
+        return new MetricName(group, typeName, name, scope, nameBuilder.toString());
+    }
+
+    public final <T> Gauge<T> newGauge(final String name, final Gauge<T> metric, final Map<String, String> tags) {
+        return KafkaYammerMetrics.defaultRegistry().newGauge(metricName(name, tags), metric);
+    }
+
+    public final <T> Gauge<T> newGauge(final String name, final Gauge<T> metric) {
+        return newGauge(name, metric, Collections.emptyMap());
+    }
+
+    public final Meter newMeter(final String name, final String eventType,
+                                final TimeUnit timeUnit, final Map<String, String> tags) {
+        return KafkaYammerMetrics.defaultRegistry().newMeter(metricName(name, tags), eventType, timeUnit);
+    }
+
+    public final Meter newMeter(final String name, final String eventType,
+                                final TimeUnit timeUnit) {
+        return newMeter(name, eventType, timeUnit, Collections.emptyMap());
+    }
+
+    public final Meter newMeter(final MetricName metricName, final String eventType, final TimeUnit timeUnit) {
+        return KafkaYammerMetrics.defaultRegistry().newMeter(metricName, eventType, timeUnit);
+    }
+
+    public final Histogram newHistogram(final String name, final boolean biased, final Map<String, String> tags) {
+        return KafkaYammerMetrics.defaultRegistry().newHistogram(metricName(name, tags), biased);
+    }
+
+    public final Histogram newHistogram(final String name) {
+        return newHistogram(name, true, Collections.emptyMap());
+    }
+
+    public final Histogram newHistogram(final String name, final boolean biased) {
+        return newHistogram(name, biased, Collections.emptyMap());
+    }
+
+    public final Histogram newHistogram(final String name, final Map<String, String> tags) {
+        return newHistogram(name, true, tags);
+    }
+
+    public final Timer newTimer(final String name, final TimeUnit durationUnit, final TimeUnit rateUnit,
+                                final Map<String, String> tags) {
+        return KafkaYammerMetrics.defaultRegistry().newTimer(metricName(name, tags), durationUnit, rateUnit);
+    }
+
+    public final Timer newTimer(final String name, final TimeUnit durationUnit, final TimeUnit rateUnit) {
+        return newTimer(name, durationUnit, rateUnit, Collections.emptyMap());
+    }
+
+    public final void removeMetric(final String name, final Map<String, String> tags) {
+        KafkaYammerMetrics.defaultRegistry().removeMetric(metricName(name, tags));
+    }
+
+    public final void removeMetric(final String name) {
+        removeMetric(name, Collections.emptyMap());
+    }
+
+    private Optional<String> toMBeanName(final Map<String, String> tags) {
```

Comment:
The ordering of `tags` matters when creating the mbean name. Unfortunately, we did not use an ordered map previously, which has resulted in unspecified behavior in some cases. When the number of tags is small, the order may have been maintained due to the Scala implementation having special implementations for thise (eg https://www.scala-lang.org/api/2.13.x/scala/collection/immutable/Set$$Set1.html).

To make this change safe, we would need to have a test that prints _all_ mbean names and verifies that they remain the same after the conversion.

Reply:
> The ordering of `tags` matters when creating the mbean name.

Does it? According to the doc [the doc](https://www.oracle.com/java/technologies/javase/management-extensions-best-practices.html#mozTocId434075), "the order of key properties is not significant". I.e. `topic=my-topic,partition=0` is equivalent to `partition=0,topic=my-topic`. AFAICS from the OpenJDK [code](https://github.com/openjdk/jdk/blob/master/src/java.management/share/classes/javax/management/ObjectName.java#L753), it's canonicalized anyway.

Reply:
Good point, my bad.